assembly/src/parsers/context.rs

```diff
@@ -285,235 +289,246 @@ impl ParserContext {
     ) -> Result<(), ParsingError> {
         while let Some(token) = tokens.read() {
             match token.parts()[0] {
+                Token::IF => nodes.push(self.parse_if(tokens)?),
                 Token::ELSE => {
                     token.validate_else()?;
                     if break_on_else {
                         break;
                     }
                     return Err(ParsingError::dangling_else(token));
                 }
-                Token::IF => {
-                    token.validate_if()?;
-                    nodes.push(self.parse_if(tokens)?);
-                }
-                Token::WHILE => {
-                    token.validate_while()?;
-                    nodes.push(self.parse_while(tokens)?);
-                }
+                Token::WHILE => nodes.push(self.parse_while(tokens)?),
                 Token::REPEAT => nodes.push(self.parse_repeat(tokens)?),
-                Token::EXEC => {
-                    let label = token.parse_exec()?;
-                    nodes.push(self.parse_exec(label, tokens)?);
-                }
-                Token::CALL => {
-                    let label = token.parse_call()?;
-                    nodes.push(self.parse_call(label, tokens)?);
-                }
-                Token::SYSCALL => {
-                    let label = token.parse_syscall()?;
-                    nodes.push(self.parse_syscall(label, tokens)?);
-                }
                 Token::END => {
                     token.validate_end()?;
                     break;
                 }
-                Token::USE | Token::EXPORT | Token::PROC | Token::BEGIN => {
-                    // TODO improve the error with the originating block
-                    // https://github.com/maticnetwork/miden/issues/514
-                    return Err(ParsingError::unexpected_body_end(token));
+                Token::USE => {
+                    return Err(ParsingError::import_inside_body(token));
+                }
+                Token::EXPORT | Token::PROC | Token::BEGIN => {
```

Comment:
Maybe add `Token::USE`; or better, delegate to a new `is_block_delimiter` function of `Token`